DEC/src/roe/RoeUtility.jl

```diff
@@ -0,0 +1,222 @@
+end
+export TypedApplication
+
+const TA = TypedApplication
+export TA
+
+Base.show(io::IO, ta::TA) = print(io, Expr(:call, nameof(ta.head), ta.sorts...))
+
+struct SortError <: Exception
+  message::String
+end
+export SortError
+
+"""   RootVar
+
+A childless node on an e-graph.
+
+"""
+@struct_hash_equal struct RootVar
+    name::Symbol
+    idx::Int
+    sort::AbstractSort
+end
+export RootVar 
+
+""" Roe
+
+Struct for storing an EGraph and its variables.
+
+Roe is the name for lobster eggs. "Egg" is the name of a Rust implementation of e-graphs, by which Metatheory.jl is inspired by. Lobsters are part of the family Decapodes, which is also the name of the AlgebraicJulia package which motivated this package. Hence, Roe.
+"""
+struct Roe
+    variables::Vector{RootVar}
+    graph::EGraph{Expr, AbstractSort}
+    function Roe()
+        new(RootVar[], EGraph{Expr, AbstractSort}())
+    end
+end
+export Roe
+
+"""
+
+A struct containing a Roe and the Id of a variable in that EGraph. The type parameter for this struct is the variable it represents.
+
+"""
+struct Var{S}
+    roe::Roe
+    id::Id
+end
+
+function EGraphs.make(g::EGraph{Expr, AbstractSort}, n::Metatheory.VecExpr)
+    op = EGraphs.get_constant(g,Metatheory.v_head(n))
+    if op isa RootVar
+        op.sort
+    elseif op isa Number
+        Scalar()
+    else
+        op((g[arg].data for arg in Metatheory.v_children(n))...)
+    end
+end
+export make
+
+function EGraphs.join(s1::AbstractSort, s2::AbstractSort)
+    if s1 == s2
+        s1
+    else
+        error("Cannot equate two nodes with different sorts")
+    end
+end
+export join
+
+function extract!(v::Var, f=EGraphs.astsize)
+    extract!(v.roe.graph, f, v.id)
+end
+export extract!
+
+function rootvarcrayon(v::RootVar)
+  lightnessrange = (50., 100.)
+  HashColor.hashcrayon(v.idx; lightnessrange, chromarange=(50., 100.))
+end
+
+function Base.show(io::IO, v::RootVar)
+  if get(io, :color, true)
+    crayon = rootvarcrayon(v)
+    print(io, crayon, "$(v.name)")
+    print(io, inv(crayon))
+  else
+    print(io, "$(v.name)#$(v.idx)")
+  end    
+end
+
+"""   fix_functions(e)::Union{Symbol, Expr}
+
+Traverses the AST of an expression, replacing the head of :call expressions to its name, a Symbol.
+"""
+function fix_functions(e)
+    @match e begin
+        s::Symbol => s
+        Expr(:call, f::Function, args...) =>
+          Expr(:call, nameof(f), fix_functions.(args)...)
+        Expr(head, args...) =>
+          Expr(head, fix_functions.(args)...)
+        _ => e
+    end
+end
+
+"""    getexpr(v::Var)::Union{Symbol, Expr}
+
+Extracts an expression (::Var) from its Roe.
+
+"""
+function getexpr(v::Var)
+    e = EGraphs.extract!(v.roe.graph, Metatheory.astsize, v.id)
+    fix_functions(e)
+end
+export getexpr
+
+function Base.show(io::IO, v::Var)
+    print(io, getexpr(v))
+end
+
+"""    fresh!(roe::Roe, sort::AbstractSort, name::Symbol)::Var{sort}
+
+Creates a new ("fresh") variable in a Roe given a sort and a name.
```

Comment:
Should be something like:

Creates a new variable in a Roe. Specifically, appends a new RootVar with given sort and name to the Roe, adds that RootVar to the e-graph, and returns a Var wrapper around the new e-graph id, with type parameter given by the sort.